# <u>Case Project - House Prices<u>

__<u>Jorge Pineño Pérez<u>__ / Case Project - House Prices

This project is of an applied nature and uses data that are available in the data file Capstone-HousePrices. The source of these data is Anglin and Gencay, “Semiparametric Estimation of a Hedonic Price Function”(Journal of Applied Econometrics 11, 1996, pages 633-648). We consider the modeling and prediction of house prices. Data are available for 546 observations of the following variables:

- `sell`: Sale price of the house
- `lot`: Lot size of the property in square feet
- `bdms`: Number of bedrooms
- `fb`: Number of full bathrooms
- `sty`: Number of stories excluding basement
- `drv`: Dummy that is 1 if the house has a driveway and 0 otherwise
- `rec`: Dummy that is 1 if the house has a recreational room and 0 otherwise
- `ffin`: Dummy that is 1 if the house has a full finished basement and 0 otherwise
- `ghw`: Dummy that is 1 if the house uses gas for hot water heating and 0 otherwise
- `ca`: Dummy that is 1 if the house uses gas for hot water heating and 0 otherwise
- `gar`: Dummy that is 1 if there is central air conditioning and 0 otherwise
- `gar`: Number of covered garage places
- `reg`: Dummy that is 1 if the house is located in a preferred neighborhood of the city and 0 otherwise
- `obs`: Observation number, needed in part (h)

In [8]:
# import packages
install.packages("readxl")
install.packages("dplyr")
install.packages("tidyr")
install.packages("lmtest")
library(readxl)
library(dplyr)
library(tidyr)
library(lmtest)

Warning message:
"package 'readxl' is in use and will not be installed"
Warning message:
"package 'dplyr' is in use and will not be installed"
Warning message:
"package 'tidyr' is in use and will not be installed"
Installing package into 'C:/Users/USER/AppData/Local/R/win-library/4.4'
(as 'lib' is unspecified)



package 'lmtest' successfully unpacked and MD5 sums checked


Warning message:
"cannot remove prior installation of package 'lmtest'"
Warning message in file.copy(savedcopy, lib, recursive = TRUE):
"problema al copiar C:\Users\USER\AppData\Local\R\win-library\4.4\00LOCK\lmtest\libs\x64\lmtest.dll  a C:\Users\USER\AppData\Local\R\win-library\4.4\lmtest\libs\x64\lmtest.dll: Permission denied"
Warning message:
"restored 'lmtest'"



The downloaded binary packages are in
	C:\Users\USER\AppData\Local\Temp\Rtmp8ccCDE\downloaded_packages


Cargando paquete requerido: zoo


Adjuntando el paquete: 'zoo'


The following objects are masked from 'package:base':

    as.Date, as.Date.numeric




In [6]:
# import the dataframe
housep <- read_excel('Case_HousePrices-round1.xls')
head(housep)

obs sell  lot  bdms fb sty drv rec ffin ghw ca gar reg
1 1   42000 5850 3    1  2   1   0   1    0   0  1   0  
2 2   38500 4000 2    1  1   1   0   0    0   0  0   0  
3 3   49500 3060 3    1  1   1   0   0    0   0  0   0  
4 4   60500 6650 3    1  2   1   1   0    0   0  0   0  
5 5   61000 6360 2    1  1   1   0   0    0   0  0   0  
6 6   66000 4160 3    1  1   1   1   1    0   1  0   0

__(a)__ Consider a linear model where the sale price of a house is the dependent variable and the explanatory variables are the other variables given above. Perform a test for linearity. What do you conclude based on the test result?

In [7]:
# regress "sell" on the other variables
price <- lm(sell ~ lot + bdms + fb + sty + drv + rec + ffin + ghw + ca + gar + reg, data=housep)
summary(price)


Call:
lm(formula = sell ~ lot + bdms + fb + sty + drv + rec + ffin + 
    ghw + ca + gar + reg, data = housep)

Residuals:
   Min     1Q Median     3Q    Max 
-41389  -9307   -591   7353  74875 

Coefficients:
              Estimate Std. Error t value Pr(>|t|)    
(Intercept) -4038.3504  3409.4713  -1.184 0.236762    
lot             3.5463     0.3503  10.124  < 2e-16 ***
bdms         1832.0035  1047.0002   1.750 0.080733 .  
fb          14335.5585  1489.9209   9.622  < 2e-16 ***
sty          6556.9457   925.2899   7.086 4.37e-12 ***
drv          6687.7789  2045.2458   3.270 0.001145 ** 
rec          4511.2838  1899.9577   2.374 0.017929 *  
ffin         5452.3855  1588.0239   3.433 0.000642 ***
ghw         12831.4063  3217.5971   3.988 7.60e-05 ***
ca          12632.8904  1555.0211   8.124 3.15e-15 ***
gar          4244.8290   840.5442   5.050 6.07e-07 ***
reg          9369.5132  1669.0907   5.614 3.19e-08 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residu

In [38]:
# perform the RESET test on our regression (adding squared and cubed fitted values)
price_reset <- resettest(price, power=2:3, type='fitted')
price_reset


	RESET test

data:  price
RESET = 13.481, df1 = 2, df2 = 532, p-value = 1.944e-06


With the RESET test, we have checked to see whether adding nonlinear transformations of the fitted values significally improved the model. The p-value which we got as a result was well below the significance level of 5%. This suggests that the model appears to be incorrectly specified, i.e. linearity is adequate.

__(b)__ Now consider a linear model where the log of the sale price of the house is the dependent variable and the explanatory variables are as before. Perform again the test for linearity. What do you conclude now?

In [24]:
# add the log of "sell" to the dataset
housep <- housep %>% mutate(logsell=log(sell))
head(housep)

obs sell  lot  bdms fb sty drv rec ffin ghw ca gar reg logsell 
1 1   42000 5850 3    1  2   1   0   1    0   0  1   0   10.64542
2 2   38500 4000 2    1  1   1   0   0    0   0  0   0   10.55841
3 3   49500 3060 3    1  1   1   0   0    0   0  0   0   10.80973
4 4   60500 6650 3    1  2   1   1   0    0   0  0   0   11.01040
5 5   61000 6360 2    1  1   1   0   0    0   0  0   0   11.01863
6 6   66000 4160 3    1  1   1   1   1    0   1  0   0   11.09741

In [26]:
# regress "logsell" on the other variables
logprice <- lm(logsell ~ lot + bdms + fb + sty + drv + rec + ffin + ghw + ca + gar + reg, data=housep)
summary(logprice)


Call:
lm(formula = logsell ~ lot + bdms + fb + sty + drv + rec + ffin + 
    ghw + ca + gar + reg, data = housep)

Residuals:
     Min       1Q   Median       3Q      Max 
-0.67865 -0.12211  0.01666  0.12868  0.67737 

Coefficients:
             Estimate Std. Error t value Pr(>|t|)    
(Intercept) 1.003e+01  4.724e-02 212.210  < 2e-16 ***
lot         5.057e-05  4.854e-06  10.418  < 2e-16 ***
bdms        3.402e-02  1.451e-02   2.345  0.01939 *  
fb          1.678e-01  2.065e-02   8.126 3.10e-15 ***
sty         9.227e-02  1.282e-02   7.197 2.10e-12 ***
drv         1.307e-01  2.834e-02   4.610 5.04e-06 ***
rec         7.352e-02  2.633e-02   2.792  0.00542 ** 
ffin        9.940e-02  2.200e-02   4.517 7.72e-06 ***
ghw         1.784e-01  4.458e-02   4.000 7.22e-05 ***
ca          1.780e-01  2.155e-02   8.262 1.14e-15 ***
gar         5.076e-02  1.165e-02   4.358 1.58e-05 ***
reg         1.271e-01  2.313e-02   5.496 6.02e-08 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' '

In [37]:
# perform the RESET test on our regression (adding squared and cubed fitted values)
logprice_reset <- resettest(logprice, power=2:3, type='fitted')
logprice_reset


	RESET test

data:  logprice
RESET = 0.13767, df1 = 2, df2 = 532, p-value = 0.8714


As we can see, the p-value which we got as a result this time was 0.87, which is this time well above the significance level of 5%. This suggests that there is less evidence suggesting that adding non-linear or ommitted variables would improve the model, and that it appears to be correctly specified, i.e. linear, as it is.

__(c)__ Continue with the linear model from question (b). Estimate a model that includes both the lot size variable and its logarithm, as well as all other explanatory variables without transformation. What is your conclusion, should we include lot size itself or its logarithm?

In [42]:
# regress "logsell" on the other variables, and the log of "lot"
logprice2 <- lm(logsell ~ lot + log(lot) + bdms + fb + sty + drv + rec + ffin + ghw + ca + gar + reg, data=housep)
summary(logprice2)


Call:
lm(formula = logsell ~ lot + log(lot) + bdms + fb + sty + drv + 
    rec + ffin + ghw + ca + gar + reg, data = housep)

Residuals:
     Min       1Q   Median       3Q      Max 
-0.68573 -0.12380  0.00785  0.12521  0.68112 

Coefficients:
              Estimate Std. Error t value Pr(>|t|)    
(Intercept)  7.150e+00  6.830e-01  10.469  < 2e-16 ***
lot         -1.490e-05  1.624e-05  -0.918 0.359086    
log(lot)     3.827e-01  9.070e-02   4.219 2.88e-05 ***
bdms         3.489e-02  1.429e-02   2.442 0.014915 *  
fb           1.659e-01  2.033e-02   8.161 2.40e-15 ***
sty          9.121e-02  1.263e-02   7.224 1.76e-12 ***
drv          1.068e-01  2.847e-02   3.752 0.000195 ***
rec          5.467e-02  2.630e-02   2.078 0.038156 *  
ffin         1.052e-01  2.171e-02   4.848 1.64e-06 ***
ghw          1.791e-01  4.390e-02   4.079 5.20e-05 ***
ca           1.643e-01  2.146e-02   7.657 9.01e-14 ***
gar          4.826e-02  1.148e-02   4.203 3.09e-05 ***
reg          1.344e-01  2.284e-02   5.88

In [40]:
# add the log of "lot" to the dataset
housep <- housep %>% mutate(loglot=log(lot))
head(housep)

obs sell  lot  bdms fb sty drv rec ffin ghw ca gar reg logsell  loglot  
1 1   42000 5850 3    1  2   1   0   1    0   0  1   0   10.64542 8.674197
2 2   38500 4000 2    1  1   1   0   0    0   0  0   0   10.55841 8.294050
3 3   49500 3060 3    1  1   1   0   0    0   0  0   0   10.80973 8.026170
4 4   60500 6650 3    1  2   1   1   0    0   0  0   0   11.01040 8.802372
5 5   61000 6360 2    1  1   1   0   0    0   0  0   0   11.01863 8.757784
6 6   66000 4160 3    1  1   1   1   1    0   1  0   0   11.09741 8.333270

After introducing the logarithm of lot size as an explanatory variable to our model, we can see how it is able to replace the variable without any transformation, and provide better results (p-value wise). This suggests that lot size does not affect house prices in a linear way, but rather exponentially.

__(d)__ Consider now a model where the log of the sale price of the house is the dependent variable and the explanatory variables are the log transformation of lot size, with all other explanatory variables as before. We now consider interaction effects of the log lot size with the other variables. Construct these interaction variables. How many are individually significant?

In [43]:
# regress "logsell" on the other variables, the log of "lot", and their interaction with it
logprice3 <- lm(logsell ~ loglot * (bdms + fb + sty + drv + rec + ffin + ghw + ca + gar + reg), data=housep)
summary(logprice3)


Call:
lm(formula = logsell ~ loglot * (bdms + fb + sty + drv + rec + 
    ffin + ghw + ca + gar + reg), data = housep)

Residuals:
     Min       1Q   Median       3Q      Max 
-0.68306 -0.11612  0.00591  0.12486  0.65998 

Coefficients:
             Estimate Std. Error t value Pr(>|t|)    
(Intercept)  8.966499   1.070667   8.375 5.09e-16 ***
loglot       0.152685   0.128294   1.190   0.2345    
bdms         0.019075   0.326700   0.058   0.9535    
fb          -0.368234   0.429048  -0.858   0.3911    
sty          0.488885   0.309700   1.579   0.1150    
drv         -1.463371   0.717225  -2.040   0.0418 *  
rec          1.673992   0.655919   2.552   0.0110 *  
ffin        -0.031844   0.445543  -0.071   0.9430    
ghw         -0.505889   0.902733  -0.560   0.5754    
ca          -0.340276   0.496041  -0.686   0.4930    
gar          0.401941   0.258646   1.554   0.1208    
reg          0.118484   0.479856   0.247   0.8051    
loglot:bdms  0.002070   0.038654   0.054   0.9573    
loglo

As we can see from our results, only two of the interaction terms would be statistically significant: the interation between the log of lot size and the driveway dummy (`loglot:drv`), and between the log of lot size and the recreational room dummy (`loglot:rec`).

__(e)__ Perform an F-test for the joint significance of the interaction effects from question (d).

In [44]:
logprice4 <- lm(logsell ~ loglot + bdms + fb + sty + drv + rec + ffin + ghw + ca + gar + reg,
                data=housep) # use only "loglot"

# create an Anova table to test tests whether all the interaction 
# terms together contribute significantly to explaining "logsell"
anova(logprice4, logprice3)

Res.Df RSS      Df Sum of Sq F        Pr(>F)   
1 534    23.63828 NA        NA       NA        NA
2 524    22.99273 10 0.6455496 1.471195 0.1465622

Since we have got as a result a p-value higher than the significance level 5% (0.15), we fail to reject the null hypothesis that all interation terms do not improve the model. Therefore, we can say that as a group, they do not significantly improve the model’s ability to explain `logsell`.

__(g)__ One may argue that some of the explanatory variables are endogenous and that there may be omitted variables. For example, the ‘condition’ of the house in terms of how it is maintained is not a variable (and difficult to measure) but will affect the house price. It will also affect, or be reflected in, some of the other variables, such as whether the house has an air conditioning (which is mostly in newer houses). If the condition of the house is missing, will the effect of air conditioning on the (log of the) sale price be over- or underestimated? (For this question no computer calculations are required.)

If the condition of the house is omitted from the model and is positively correlated with both the presence of air conditioning and the house price, then the estimated effect of air conditioning on the log of the sale price will be overestimated. This is because the regression will attribute to air conditioning not only its own effect but also part of the positive effect of house condition (omitted variable bias). Since better-maintained or newer homes are more likely to have air conditioning and also tend to sell for higher prices, the unobserved “condition” variable inflates the apparent impact of air conditioning on price.

__(h)__ Finally we analyze the predictive ability of the model. Consider again the model where the log of the sale price of the house is the dependent variable and the explanatory variables are the log transformation of lot size, with all other explanatory variables in their original form (and no interaction effects). Estimate the parameters of the model using the first 400 observations. Make predictions on the log of the price and calculate the MAE for the other 146 observations. How good is the predictive power of the model (relative to the variability in the log of the price)?

In [45]:
# divide the dataset between the estimation and testing observations
estim <- housep[1:400, ]
test <- housep[401:546, ]

# estimate the model using only the 400 first observations
logprice_estim <- lm(logsell ~ loglot + bdms + fb + sty + drv + rec + ffin + ghw + ca + gar + reg,
                     data=estim)

# make predictions for the 146 other observations
logsell_pred <- predict(logprice_estim, newdata=test)

# compute MAE
logsell_actual <- test$logsell
mae <- mean(abs(logsell_pred - logsell_actual))
mae

[1] 0.1278416

In [46]:
# get the standard deviation of "logsell" in the test set
logsell_sd <- sd(logsell_actual)
logsell_sd

[1] 0.2887723

With a mean absolute error (MAE) of 0.13 and a standard deviation of 0.29 in the log of the sale price, the model demonstrates reasonably good predictive power. Since the MAE is less than half the standard deviation, the prediction errors are relatively small compared to the natural variability in the data. This indicates that, althouth not perfect, the model is able to capture a substantial portion of the variation in house prices and provides a meaningful improvement over simple predictions based on the average price alone.